src/backend/distributed/commands/schema.c

```diff
@@ -296,3 +320,101 @@ EnsureSequentialModeForSchemaDDL(void)
 							   "use only one connection for all future commands")));
 	SetLocalMultiShardModifyModeToSequential();
 }
+
+
+static bool
+NameListHasDistributedSchema(List *schemaNames)
+{
+	Value *schemaVal;
+	foreach_ptr(schemaVal, schemaNames)
+	{
+		ObjectAddress address = GetObjectAddressBySchemaName(strVal(schemaVal), true);
+
+		if (IsObjectDistributed(&address))
+		{
+			return true;
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * InvalidateFKeysForDistributedTablesInSchema takes a schema name and marks
+ * foreign graph to be invalidated, if a distributed table in the given schema
+ * has a foreign key connection.
+ */
+static void
+InvalidateFKeysForDistributedTablesInSchema(char *schemaName)
+{
+	Relation pgClass = NULL;
+	HeapTuple heapTuple = NULL;
+	SysScanDesc scanDescriptor = NULL;
+	ScanKeyData scanKey[1];
+	int scanKeyCount = 1;
+	Oid scanIndexId = InvalidOid;
+	bool useIndex = false;
+
+	Oid namespaceOid = get_namespace_oid(schemaName, true);
+
+	if (namespaceOid == InvalidOid)
+	{
+		return;
+	}
+
+	pgClass = table_open(RelationRelationId, AccessShareLock);
+	
+	ScanKeyInit(&scanKey[0], Anum_pg_class_relnamespace, BTEqualStrategyNumber,
+				F_OIDEQ, namespaceOid);
+	scanDescriptor = systable_beginscan(pgClass, scanIndexId, useIndex, NULL,
+										scanKeyCount, scanKey);
+
+	heapTuple = systable_getnext(scanDescriptor);
+	while (HeapTupleIsValid(heapTuple))
+	{
+		Form_pg_class relationForm = (Form_pg_class) GETSTRUCT(heapTuple);
+		char *relationName = NameStr(relationForm->relname);
+		Oid relationId = get_relname_relid(relationName, namespaceOid);
+
+		/* we're not interested in non-valid, non-distributed relations */
+		if (relationId == InvalidOid || !IsCitusTable(relationId))
+		{
+			heapTuple = systable_getnext(scanDescriptor);
+			continue;
+		}
+
+		/* invalidate foreign key cache if the table involved in any foreign key */
+		if (TableReferenced(relationId) || TableReferencing(relationId))
+		{
+			MarkInvalidateForeignKeyGraph();
+		}
+
+		heapTuple = systable_getnext(scanDescriptor);
+	}
+
+	systable_endscan(scanDescriptor);
+	table_close(pgClass, NoLock);
+}
+
+
+static bool
+ShouldPropagateCreateSchemaStmt()
+{
+	if (!ShouldPropagate())
+	{
+		return false;
+	}
+
+	if (!IsMultiStatementTransaction())
+	{
+		return true;
+	}
+
+	if (MultiShardConnectionType != SEQUENTIAL_CONNECTION)
+	{
+		return false;
+	}
```

Comment:
These two statements are highly correlated, I would prefer to have a positive check on the transaction with the sequential check embedded in there to make the logic slightly easier to grasp.

The combination of negating if statements with alternating return values make it _non_ trivial to grasp the logic.

Nevertheless, this could use a bit of comments explaining what we are checking and why.

```suggestion
	if (IsMultiStatementTransaction())
	{
		if (MultiShardConnectionType != SEQUENTIAL_CONNECTION)
		{
			return false;
		}
	}
```